src/webgpu/shader/execution/statement/discard.spec.ts

```diff
@@ -0,0 +1,606 @@
+          ],
+        }
+      );
+    };
+
+    // No fragment outputs occur.
+    const fbChecker = (a: Uint32Array) => {
+      return checkElementsPassPredicate(
+        a,
+        (idx: number, value: number | bigint) => {
+          return value === 0;
+        },
+        {
+          predicatePrinter: [
+            {
+              leftHeader: 'fb exp ==',
+              getValueForCell: (idx: number) => {
+                return 0;
+              },
+            },
+          ],
+        }
+      );
+    };
+
+    drawFullScreen(t, code, dataChecker, fbChecker);
+  });
+
+g.test('three_quarters')
+  .desc('Test a shader that discards all fragments')
+  .fn(t => {
+    const code = `
+${kSharedCode}
+
+@fragment
+fn fsMain(@builtin(position) pos : vec4f) -> @location(0) u32 {
+  _ = uniformValues[0];
+  if (pos.x >= 0.5 || pos.y >= 0.5) {
+    discard;
+  }
+  let idx = atomicAdd(&atomicIndex, 1);
+  output[idx] = pos.xy;
+  return idx;
+}
+`;
+
+    // Only the the upper left quadrant is kept.
+    const dataChecker = (a: Float32Array) => {
+      return checkElementsPassPredicate(
+        a,
+        (idx: number, value: number | bigint) => {
+          return value < 0.5;
+        },
+        {
+          predicatePrinter: [
+            {
+              leftHeader: 'data exp ==',
+              getValueForCell: (idx: number): number | string => {
+                const is_x = idx % 2 === 0;
+                if (is_x) {
+                  const x = Math.floor(idx / 2) % kWidth;
+                  if (x >= kWidth / 2) {
+                    return 0;
+                  }
+                } else {
+                  const y = Math.floor((idx - 1) / kWidth);
+                  if (y >= kHeight / 2) {
+                    return 0;
+                  }
+                }
+                return '< 0.5';
+              },
+            },
+          ],
+        }
+      );
+    };
+    const fbChecker = (a: Uint32Array) => {
+      return checkElementsPassPredicate(
+        a,
+        (idx: number, value: number | bigint) => {
+          return value < (kWidth * kHeight) / 4 && (value === a[idx - 1] || value === 0);
+        },
+        {
+          predicatePrinter: [
+            {
+              leftHeader: 'fb exp ==',
+              getValueForCell: (idx: number) => {
+                const x = idx % kWidth;
+                const y = Math.floor(idx / kWidth);
+                if (x < kWidth / 2 && y < kHeight / 2) {
+                  return 'any';
+                } else {
+                  return 0;
+                }
+              },
+            },
+          ],
+        }
+      );
+    };
+
+    drawFullScreen(t, code, dataChecker, fbChecker);
+  });
+
+g.test('function_call')
+  .desc('Test discards happening in a function call')
+  .fn(t => {
+    const code = `
+${kSharedCode}
+
+fn foo(pos : vec2f) {
+  if (pos.x <= 0.5 && pos.y <= 0.5) {
+    discard;
+  }
+  if (pos.x >= 0.5 && pos.y >= 0.5) {
+    discard;
+  }
```

Comment:
Again, I think these position checks are off, or the predicates aren't robust (or both). I can turn this into an unconditional discard and the test still passes.